elpis/engines/espnet/objects/model.py

```diff
@@ -0,0 +1,146 @@
+from elpis.engines.common.objects.model import Model as BaseModel
+from elpis.engines.kaldi.input.json_to_kaldi import create_kaldi_structure
+
+
+class EspnetModel(BaseModel):
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        # ESPnet does not use a pronunciation dictionary so this will not
+        # change from None.
+        self.pron_dict = None
+        self.config['pron_dict_name'] = None
+        # ESPnet doesn't use an n-gram language model, so this will not change
+        # from None.
+        self.config['ngram'] = None
+
+    @classmethod
+    def load(cls, base_path: Path):
+        self = super().load(base_path)
+        self.pron_dict = None
+        return self
+
+    @property
+    def status(self):
+        return self.config['status']
+
+    @property
+    def state(self):
+        # In the KaldiModel that I'm basing this code of there is a "TODO: fix
+        # this" in this method. I'm not sure what that's about.
+        return {}
+
+    def has_been_trained(self):
+        return self.status == 'trained'
+
+    @status.setter
+    def status(self, value: str):
+        self.config['status'] = value
+
+    def link(self, dataset: Dataset, _pron_dict):
+        self.dataset = dataset
+        self.config['dataset_name'] = dataset.name
+        # Note the _pron_dict is ignored as it's irrelevant to ESPnet.
+
+    def build_structure(self):
+        print("BUILD STRUCTURE")
+        # NOTE Since the ESPnet data is similari informatting requirements to
+        # Kaldi structure, code is unfortunately being duplicated from KaldiModel.
+        # I'm not sure the best way to get around this, but calling create_kaldi_structure()
+        # here does limit that duplication somewhat.
+        output_path = self.path.joinpath('output')
+        output_path.mkdir(parents=True, exist_ok=True)
+
+        print("OUTPUT PATH: {}".format(output_path))
+
+        # Copy cleaned corpus from dataset to the model
+        dataset_corpus_txt = self.dataset.path.joinpath('cleaned', 'corpus.txt')
+        model_corpus_txt = self.path.joinpath('corpus.txt')
+        if os.path.exists(dataset_corpus_txt):
+            shutil.copy(f'{dataset_corpus_txt}', f'{model_corpus_txt}')
+        create_kaldi_structure(
+            input_json=f'{self.dataset.pathto.annotation_json}',
+            output_folder=f'{output_path}',
+            silence_markers=False,
+            corpus_txt=f'{model_corpus_txt}'
+        )
+
+    def train(self, on_complete:Callable=None):
+
+        def prepare_for_training():
+            # This should just copy the ESPnet experiment directory into the
+            # model directory and then copy the prepared train/test langdir
+            # stuff into the appropriate subdirectory.
+
+            # First make a copy of the ESPNET Elpis recipe
+            model_path = Path(self.path)
+            local_espnet_path = model_path / "espnet-asr1"
```

Comment:
Would be great to have consistency of use between `Path() / addition` and `Path().joinpath()`, at least within this module. The rest of the project tends to use `.joinpath()`.